File: tests/algorithms/x/Java/dynamic_programming/abbreviation.java
public class Main {

    static int index_of(String s, String ch) {
        int i = 0;
        while (i < _runeLen(s)) {
            if ((s.substring(i, i+1).equals(ch))) {
                return i;
            }
            i = i + 1;
        }
        return -1;
    }

    static int ord(String ch) {
        String upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        String lower = "abcdefghijklmnopqrstuvwxyz";
        int idx = index_of(upper, ch);
        if (idx >= 0) {
            return 65 + idx;
        }
        idx = index_of(lower, ch);
        if (idx >= 0) {
            return 97 + idx;
        }
        return 0;
    }

    static String chr(int n) {
        String upper_1 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        String lower_1 = "abcdefghijklmnopqrstuvwxyz";
        if (n >= 65 && n < 91) {
            return upper_1.substring(n - 65, n - 64);
        }
        if (n >= 97 && n < 123) {
            return lower_1.substring(n - 97, n - 96);
        }
        return "?";
    }

    static String to_upper_char(String c) {
        int code = ord(c);
        if (code >= 97 && code <= 122) {
            return chr(code - 32);
        }
        return c;
    }

    static boolean is_lower(String c) {
        int code_1 = ord(c);
        return code_1 >= 97 && code_1 <= 122;
    }

    static boolean abbr(String a, String b) {
        int n = _runeLen(a);
        int m = _runeLen(b);
        boolean[][] dp = ((boolean[][])(new boolean[][]{}));
        int i_1 = 0;
        while (i_1 <= n) {
            boolean[] row = ((boolean[])(new boolean[]{}));
            int j = 0;
            while (j <= m) {
                row = ((boolean[])(appendBool(row, false)));
                j = j + 1;
            }
            dp = ((boolean[][])(appendObj(dp, row)));
            i_1 = i_1 + 1;
        }
dp[0][0] = true;
        i_1 = 0;
        while (i_1 < n) {
            int j_1 = 0;
            while (j_1 <= m) {
                if (((Boolean)(dp[i_1][j_1]))) {
                    if (j_1 < m && (to_upper_char(a.substring(i_1, i_1+1)).equals(b.substring(j_1, j_1+1)))) {
dp[i_1 + 1][j_1 + 1] = true;
                    }
                    if (((Boolean)(is_lower(a.substring(i_1, i_1+1))))) {
dp[i_1 + 1][j_1] = true;
                    }
                }
                j_1 = j_1 + 1;
            }
            i_1 = i_1 + 1;
        }
        return dp[n][m];
    }

    static void print_bool(boolean b) {
        if (((Boolean)(b))) {
            System.out.println(true ? "True" : "False");
        } else {
            System.out.println(false ? "True" : "False");
        }
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            print_bool(abbr("daBcd", "ABC"));
            print_bool(abbr("dBcd", "ABC"));
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static boolean[] appendBool(boolean[] arr, boolean v) {
        boolean[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }

    static <T> T[] appendObj(T[] arr, T v) {
        T[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }
}

File: tests/algorithms/transpiler/FS/dynamic_programming/abbreviation.fs
// Generated 2025-08-07 15:46 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec index_of (s: string) (ch: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable s = s
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (String.length (s)) do
            if (string (s.[i])) = ch then
                __ret <- i
                raise Return
            i <- i + 1
        __ret <- -1
        raise Return
        __ret
    with
        | Return -> __ret
let rec ord (ch: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable ch = ch
    try
        let upper: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        let lower: string = "abcdefghijklmnopqrstuvwxyz"
        let mutable idx: int = index_of (upper) (ch)
        if idx >= 0 then
            __ret <- 65 + idx
            raise Return
        idx <- index_of (lower) (ch)
        if idx >= 0 then
            __ret <- 97 + idx
            raise Return
        __ret <- 0
        raise Return
        __ret
    with
        | Return -> __ret
let rec chr (n: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable n = n
    try
        let upper: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        let lower: string = "abcdefghijklmnopqrstuvwxyz"
        if (n >= 65) && (n < 91) then
            __ret <- upper.Substring(n - 65, (n - 64) - (n - 65))
            raise Return
        if (n >= 97) && (n < 123) then
            __ret <- lower.Substring(n - 97, (n - 96) - (n - 97))
            raise Return
        __ret <- "?"
        raise Return
        __ret
    with
        | Return -> __ret
let rec to_upper_char (c: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable c = c
    try
        let code: int = ord (c)
        if (code >= 97) && (code <= 122) then
            __ret <- chr (code - 32)
            raise Return
        __ret <- c
        raise Return
        __ret
    with
        | Return -> __ret
let rec is_lower (c: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable c = c
    try
        let code: int = ord (c)
        __ret <- (code >= 97) && (code <= 122)
        raise Return
        __ret
    with
        | Return -> __ret
let rec abbr (a: string) (b: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable a = a
    let mutable b = b
    try
        let n: int = String.length (a)
        let m: int = String.length (b)
        let mutable dp: bool array array = [||]
        let mutable i: int = 0
        while i <= n do
            let mutable row: bool array = [||]
            let mutable j: int = 0
            while j <= m do
                row <- Array.append row [|false|]
                j <- j + 1
            dp <- Array.append dp [|row|]
            i <- i + 1
        dp.[0].[0] <- true
        i <- 0
        while i < n do
            let mutable j: int = 0
            while j <= m do
                if _idx (_idx dp (i)) (j) then
                    if (j < m) && ((to_upper_char (string (a.[i]))) = (string (b.[j]))) then
                        dp.[i + 1].[j + 1] <- true
                    if is_lower (string (a.[i])) then
                        dp.[i + 1].[j] <- true
                j <- j + 1
            i <- i + 1
        __ret <- _idx (_idx dp (n)) (m)
        raise Return
        __ret
    with
        | Return -> __ret
let rec print_bool (b: bool) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable b = b
    try
        if b then
            printfn "%b" (true)
        else
            printfn "%b" (false)
        __ret
    with
        | Return -> __ret
print_bool (abbr ("daBcd") ("ABC"))
print_bool (abbr ("dBcd") ("ABC"))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
